From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Carlo=20Maria=20Card=C3=AD?=
 <112445885+MathsAnalysis@users.noreply.github.com>
Date: Thu, 4 Jun 2026 09:35:46 +0200
Subject: [PATCH] MSPT Tracking for each world


diff --git a/net/minecraft/server/MinecraftServer.java b/net/minecraft/server/MinecraftServer.java
index 5c587d5120ae4b7120de5aad135b446af8b1ab24..b32dc47b20a462dc12d54c60c7e454dea3165f21 100644
--- a/net/minecraft/server/MinecraftServer.java
+++ b/net/minecraft/server/MinecraftServer.java
@@ -1885,6 +1885,8 @@ public abstract class MinecraftServer extends ReentrantBlockableEventLoop<TickTa
             profiler.push(() -> level + " " + level.dimension().identifier());
             profiler.push("tick");
 
+            // Purpur start - per-world MSPT tracking (DivineMC)
+            final long worldTickStart = Util.getNanos();
             try {
                 level.tick(haveTime);
             } catch (Throwable t) {
@@ -1892,6 +1894,11 @@ public abstract class MinecraftServer extends ReentrantBlockableEventLoop<TickTa
                 level.fillReportDetails(report);
                 throw new ReportedException(report);
             }
+            final long worldTickDuration = Util.getNanos() - worldTickStart;
+            level.tickTimes5s.add(this.tickCount, worldTickDuration);
+            level.tickTimes10s.add(this.tickCount, worldTickDuration);
+            level.tickTimes60s.add(this.tickCount, worldTickDuration);
+            // Purpur end - per-world MSPT tracking (DivineMC)
 
             profiler.pop();
             profiler.pop();
diff --git a/net/minecraft/server/level/ServerLevel.java b/net/minecraft/server/level/ServerLevel.java
index d156b872ed38f19df15290f07ab1b3866c14574f..0c64ed9c2ccebabe05e89dca9d0d82173cb93636 100644
--- a/net/minecraft/server/level/ServerLevel.java
+++ b/net/minecraft/server/level/ServerLevel.java
@@ -607,6 +607,12 @@ public class ServerLevel extends Level implements WorldGenLevel, ServerEntityGet
     }
     // Paper end - chunk tick iteration
 
+    // Purpur start - per-world MSPT tracking (DivineMC)
+    public final MinecraftServer.TickTimes tickTimes5s = new MinecraftServer.TickTimes(100);
+    public final MinecraftServer.TickTimes tickTimes10s = new MinecraftServer.TickTimes(200);
+    public final MinecraftServer.TickTimes tickTimes60s = new MinecraftServer.TickTimes(1200);
+    // Purpur end - per-world MSPT tracking (DivineMC)
+
     public ServerLevel(
         final MinecraftServer server,
         final Executor executor,
