From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Carlo=20Maria=20Card=C3=AD?=
 <112445885+MathsAnalysis@users.noreply.github.com>
Date: Thu, 4 Jun 2026 02:48:54 +0200
Subject: [PATCH] Petal Multithreaded Tracker


diff --git a/ca/spottedleaf/moonrise/common/misc/NearbyPlayers.java b/ca/spottedleaf/moonrise/common/misc/NearbyPlayers.java
index fff7a796a61542213bdaa417d44e25cd05f7e774..39c8816068181dd266a47d20cf0617b322ba65ad 100644
--- a/ca/spottedleaf/moonrise/common/misc/NearbyPlayers.java
+++ b/ca/spottedleaf/moonrise/common/misc/NearbyPlayers.java
@@ -60,7 +60,16 @@ public final class NearbyPlayers {
 
     private final ServerLevel world;
     private final Reference2ReferenceOpenHashMap<ServerPlayer, TrackedPlayer[]> players = new Reference2ReferenceOpenHashMap<>();
-    private final Long2ReferenceOpenHashMap<TrackedChunk> byChunk = new Long2ReferenceOpenHashMap<>();
+    // Purpur start - async entity tracker (DivineMC)
+    private final it.unimi.dsi.fastutil.longs.Long2ReferenceMap<TrackedChunk> byChunk;
+    {
+        if (org.bxteam.divinemc.config.DivineConfig.AsyncCategory.multithreadedEnabled) {
+            byChunk = it.unimi.dsi.fastutil.longs.Long2ReferenceMaps.synchronize(new Long2ReferenceOpenHashMap<>());
+        } else {
+            byChunk = new Long2ReferenceOpenHashMap<>();
+        }
+    }
+    // Purpur end - async entity tracker (DivineMC)
     private final Long2ReferenceOpenHashMap<ReferenceList<ServerPlayer>>[] directByChunk = new Long2ReferenceOpenHashMap[TOTAL_MAP_TYPES];
     {
         for (int i = 0; i < this.directByChunk.length; ++i) {
diff --git a/ca/spottedleaf/moonrise/patches/chunk_system/player/RegionizedPlayerChunkLoader.java b/ca/spottedleaf/moonrise/patches/chunk_system/player/RegionizedPlayerChunkLoader.java
index e3e9ee76e943d95293f2c2ec64a610d56bc05491..4a9a8732c364dc98d2a713a3cf35aa36ef947a45 100644
--- a/ca/spottedleaf/moonrise/patches/chunk_system/player/RegionizedPlayerChunkLoader.java
+++ b/ca/spottedleaf/moonrise/patches/chunk_system/player/RegionizedPlayerChunkLoader.java
@@ -344,8 +344,11 @@ public final class RegionizedPlayerChunkLoader {
         private boolean canGenerateChunks = true;
 
         private final ArrayDeque<ChunkHolderManager.TicketOperation<?, ?>> delayedTicketOps = new ArrayDeque<>();
-        private final LongOpenHashSet sentChunks = new LongOpenHashSet();
-
+        // Purpur start - async entity tracker (DivineMC)
+        private final LongOpenHashSet sentChunks = org.bxteam.divinemc.config.DivineConfig.AsyncCategory.multithreadedEnabled && !org.bxteam.divinemc.config.DivineConfig.AsyncCategory.multithreadedCompatModeEnabled
+            ? new org.bxteam.divinemc.util.map.ConcurrentLongHashSet()
+            : new LongOpenHashSet();
+        // Purpur end - async entity tracker (DivineMC)
         private static final byte CHUNK_TICKET_STAGE_NONE           = 0;
         private static final byte CHUNK_TICKET_STAGE_LOADING        = 1;
         private static final byte CHUNK_TICKET_STAGE_LOADED         = 2;
diff --git a/net/minecraft/network/protocol/game/ClientboundUpdateAttributesPacket.java b/net/minecraft/network/protocol/game/ClientboundUpdateAttributesPacket.java
index 344260d191e4c80671062cd29ce0146a16ea7f1d..7507fd9fbf63360aef45430d4eb5cfc3df3a2656 100644
--- a/net/minecraft/network/protocol/game/ClientboundUpdateAttributesPacket.java
+++ b/net/minecraft/network/protocol/game/ClientboundUpdateAttributesPacket.java
@@ -32,6 +32,7 @@ public class ClientboundUpdateAttributesPacket implements Packet<ClientGamePacke
         this.attributes = Lists.newArrayList();
 
         for (AttributeInstance value : values) {
+            if (value == null) continue; // Purpur - async entity tracker (DivineMC)
             this.attributes.add(new ClientboundUpdateAttributesPacket.AttributeSnapshot(value.getAttribute(), value.getBaseValue(), value.getModifiers()));
         }
     }
diff --git a/net/minecraft/server/level/ChunkMap.java b/net/minecraft/server/level/ChunkMap.java
index 42ec67c53c35189838ce4767c2be620e05108159..c44215ab2d4b52ddb496124b74abf6d7424b4cf9 100644
--- a/net/minecraft/server/level/ChunkMap.java
+++ b/net/minecraft/server/level/ChunkMap.java
@@ -252,9 +252,19 @@ public class ChunkMap extends SimpleRegionStorage implements ChunkHolder.PlayerP
         }
 
         final ServerPlayer[] backingSet = inRange.getRawDataUnchecked();
-        for (int i = 0, len = inRange.size(); i < len; i++) {
-            ++(backingSet[i].mobCounts[index]);
+        // Purpur start - async entity tracker (DivineMC)
+        if (org.bxteam.divinemc.config.DivineConfig.AsyncCategory.multithreadedEnabled) {
+            for (int i = 0, len = inRange.size(); i < len; i++) {
+                final ServerPlayer player = backingSet[i];
+                if (player == null) continue;
+                ++(player.mobCounts[index]);
+            }
+        } else {
+            for (int i = 0, len = inRange.size(); i < len; i++) {
+                ++(backingSet[i].mobCounts[index]);
+            }
         }
+        // Purpur end - async entity tracker (DivineMC)
     }
 
     // Paper start - per player mob count backoff
@@ -1057,6 +1067,13 @@ public class ChunkMap extends SimpleRegionStorage implements ChunkHolder.PlayerP
     // Paper end - optimise entity tracker
 
     protected void tick() {
+        // Purpur start - async entity tracker (DivineMC)
+        if (org.bxteam.divinemc.config.DivineConfig.AsyncCategory.multithreadedEnabled) {
+            final ServerLevel level = this.level;
+            org.bxteam.divinemc.async.tracking.MultithreadedTracker.tick(level);
+            return;
+        }
+        // Purpur end - async entity tracker (DivineMC)
         // Paper start - optimise entity tracker
         if (true) {
             this.newTrackerTick();
@@ -1208,8 +1225,44 @@ public class ChunkMap extends SimpleRegionStorage implements ChunkHolder.PlayerP
         private final Entity entity;
         private final int range;
         private SectionPos lastSectionPos;
-        public final Set<ServerPlayerConnection> seenBy = new it.unimi.dsi.fastutil.objects.ReferenceOpenHashSet<>(); // Paper - Perf: optimise map impl
+        // Purpur start - async entity tracker (DivineMC)
+        public static final ServerPlayerConnection[] EMPTY_OBJECT_ARRAY = new ServerPlayerConnection[0];
+        private final it.unimi.dsi.fastutil.objects.ReferenceOpenHashSet<ServerPlayerConnection> nonSyncSeenBy = new it.unimi.dsi.fastutil.objects.ReferenceOpenHashSet<>() {
+            @Override
+            public boolean add(ServerPlayerConnection serverPlayerConnection) {
+                seenByUpdated = true;
+                return super.add(serverPlayerConnection);
+            }
+
+            @Override
+            public boolean remove(Object k) {
+                seenByUpdated = true;
+                return super.remove(k);
+            }
+
+            @Override
+            public void clear() {
+                seenByUpdated = true;
+                super.clear();
+            }
+        };
+        public final Set<ServerPlayerConnection> seenBy = org.bxteam.divinemc.config.DivineConfig.AsyncCategory.multithreadedEnabled ? it.unimi.dsi.fastutil.objects.ReferenceSets.synchronize(nonSyncSeenBy) : nonSyncSeenBy; // Paper - Perf: optimise map impl
+        private volatile boolean seenByUpdated = true;
+        private volatile ServerPlayerConnection[] seenByArray = EMPTY_OBJECT_ARRAY;
+
+        public ServerPlayerConnection[] seenBy() {
+            if (!seenByUpdated) {
+                return seenByArray;
+            } else {
+                return seenBy.toArray(EMPTY_OBJECT_ARRAY);
+            }
+        }
 
+        public void seenByUpdated() {
+            this.seenByArray = this.seenBy.toArray(EMPTY_OBJECT_ARRAY);
+            seenByUpdated = false;
+        }
+        // Purpur end - async entity tracker (DivineMC)
         // Paper start - optimise entity tracker
         private long lastChunkUpdate = -1L;
         private ca.spottedleaf.moonrise.common.misc.NearbyPlayers.TrackedChunk lastTrackedChunk;
@@ -1235,23 +1288,92 @@ public class ChunkMap extends SimpleRegionStorage implements ChunkHolder.PlayerP
             this.lastTrackedChunk = chunk;
 
             final ServerPlayer[] playersRaw = players.getRawDataUnchecked();
+            final int playersLength = Math.min(playersRaw.length, players.size()); // Purpur - async entity tracker (DivineMC)
 
-            for (int i = 0, len = players.size(); i < len; ++i) {
+            for (int i = 0; i < playersLength; ++i) { // Purpur - async entity tracker (DivineMC)
                 final ServerPlayer player = playersRaw[i];
                 this.updatePlayer(player);
             }
 
             if (lastChunkUpdate != currChunkUpdate || lastTrackedChunk != chunk) {
                 // need to purge any players possible not in the chunk list
-                for (final ServerPlayerConnection conn : new java.util.ArrayList<>(this.seenBy)) {
+                // Purpur start - async entity tracker (DivineMC)
+                boolean removed = false;
+                for (final ServerPlayerConnection conn : this.seenBy()) {
                     final ServerPlayer player = conn.getPlayer();
                     if (!players.contains(player)) {
-                        this.removePlayer(player);
+                        removed |= this.removePlayerMulti(player);
                     }
                 }
+
+                if (removed) {
+                    this.seenByUpdated();
+                }
+                // Purpur end - async entity tracker (DivineMC)
             }
         }
 
+        // Purpur start - async entity tracker (DivineMC)
+        public final @Nullable Runnable tickCompact(final ca.spottedleaf.moonrise.common.misc.NearbyPlayers.TrackedChunk chunk) {
+            if (chunk == null) {
+                this.moonrise$clearPlayers();
+                return null;
+            }
+
+            final ca.spottedleaf.moonrise.common.list.ReferenceList<ServerPlayer> players = chunk.getPlayers(ca.spottedleaf.moonrise.common.misc.NearbyPlayers.NearbyMapType.VIEW_DISTANCE);
+
+            if (players == null) {
+                this.moonrise$clearPlayers();
+                return null;
+            }
+
+            final long lastChunkUpdate = this.lastChunkUpdate;
+            final long currChunkUpdate = chunk.getUpdateCount();
+            final ca.spottedleaf.moonrise.common.misc.NearbyPlayers.TrackedChunk lastTrackedChunk = this.lastTrackedChunk;
+            this.lastChunkUpdate = currChunkUpdate;
+            this.lastTrackedChunk = chunk;
+
+            final ServerPlayer[] playersRaw = players.getRawDataUnchecked();
+            final int playersLen = players.size(); // Ensure length won't change in the future tasks
+
+            if (!org.bxteam.divinemc.config.DivineConfig.AsyncCategory.multithreadedEnabled || !org.bxteam.divinemc.config.DivineConfig.AsyncCategory.multithreadedCompatModeEnabled) {
+                throw new IllegalStateException();
+            }
+            final boolean isServerPlayer = this.entity instanceof ServerPlayer;
+            final boolean isRealPlayer = isServerPlayer && ((ca.spottedleaf.moonrise.patches.chunk_system.player.ChunkSystemServerPlayer) this.entity).moonrise$isRealPlayer();
+            Runnable updatePlayerTasks = () -> {
+                for (int i = 0; i < playersLen; ++i) {
+                    final ServerPlayer player = playersRaw[i];
+                    this.updatePlayer(player);
+                }
+
+                if (lastChunkUpdate != currChunkUpdate || lastTrackedChunk != chunk) {
+                    // need to purge any players possible not in the chunk list
+                    boolean removed = false;
+                    for (final ServerPlayerConnection conn : this.seenBy()) {
+                        final ServerPlayer player = conn.getPlayer();
+                        if (!players.contains(player)) {
+                            removed |= this.removePlayerMulti(player);
+                        }
+                    }
+                    if (removed) {
+                        this.seenByUpdated();
+                    }
+                }
+            };
+
+            // Only update asynchronously for real player, and sync update for fake players
+            // This can fix compatibility issue with NPC plugins using real entity type, like Citizens
+            // To prevent visible issue with player type NPCs
+            // btw, still recommend to use packet based NPC plugins, like ZNPC Plus, Adyeshach, Fancy NPC, etc.
+            if (isRealPlayer || !isServerPlayer) {
+                return updatePlayerTasks;
+            } else {
+                updatePlayerTasks.run();
+                return null;
+            }
+        }
+        // Purpur end - async entity tracker (DivineMC)
         @Override
         public final void moonrise$removeNonTickThreadPlayers() {
             boolean foundToRemove = false;
@@ -1266,12 +1388,13 @@ public class ChunkMap extends SimpleRegionStorage implements ChunkHolder.PlayerP
                 return;
             }
 
-            for (final ServerPlayerConnection conn : new java.util.ArrayList<>(this.seenBy)) {
+            for (final ServerPlayerConnection conn : this.seenBy()) { // Purpur - async entity tracker (DivineMC)
                 ServerPlayer player = conn.getPlayer();
                 if (!ca.spottedleaf.moonrise.common.util.TickThread.isTickThreadFor(player)) {
-                    this.removePlayer(player);
+                    this.removePlayerMulti(player); // Purpur - async entity tracker (DivineMC)
                 }
             }
+            this.seenByUpdated(); // Purpur - async entity tracker (DivineMC)
         }
 
         @Override
@@ -1281,10 +1404,11 @@ public class ChunkMap extends SimpleRegionStorage implements ChunkHolder.PlayerP
             if (this.seenBy.isEmpty()) {
                 return;
             }
-            for (final ServerPlayerConnection conn : new java.util.ArrayList<>(this.seenBy)) {
+            for (final ServerPlayerConnection conn : this.seenBy()) { // Purpur - async entity tracker (DivineMC)
                 ServerPlayer player = conn.getPlayer();
-                this.removePlayer(player);
+                this.removePlayerMulti(player); // Purpur - async entity tracker (DivineMC)
             }
+            this.seenByUpdated(); // Purpur - async entity tracker (DivineMC)
         }
 
         @Override
@@ -1312,7 +1436,7 @@ public class ChunkMap extends SimpleRegionStorage implements ChunkHolder.PlayerP
 
         @Override
         public void sendToTrackingPlayers(final Packet<? super ClientGamePacketListener> packet) {
-            for (ServerPlayerConnection connection : this.seenBy) {
+            for (ServerPlayerConnection connection : this.seenBy()) { // Purpur - async entity tracker (DivineMC)
                 connection.send(packet);
             }
         }
@@ -1327,7 +1451,7 @@ public class ChunkMap extends SimpleRegionStorage implements ChunkHolder.PlayerP
 
         @Override
         public void sendToTrackingPlayersFiltered(final Packet<? super ClientGamePacketListener> packet, final Predicate<ServerPlayer> targetPredicate) {
-            for (ServerPlayerConnection connection : this.seenBy) {
+            for (ServerPlayerConnection connection : this.seenBy()) { // Purpur - async entity tracker (DivineMC)
                 if (targetPredicate.test(connection.getPlayer())) {
                     connection.send(packet);
                 }
@@ -1335,24 +1459,34 @@ public class ChunkMap extends SimpleRegionStorage implements ChunkHolder.PlayerP
         }
 
         public void broadcastRemoved() {
-            for (ServerPlayerConnection connection : this.seenBy) {
+            for (ServerPlayerConnection connection : this.seenBy()) { // Purpur - async entity tracker (DivineMC)
                 this.serverEntity.removePairing(connection.getPlayer());
             }
         }
 
+        // Purpur start - async entity tracker (DivineMC)
+        public boolean removePlayerMulti(ServerPlayer player) {
+            if (this.seenBy.remove(player.connection)) {
+                this.serverEntity.removePairing(player);
+                return true;
+            } else {
+                return false;
+            }
+        }
+        // Purpur end - async entity tracker (DivineMC)
         public void removePlayer(final ServerPlayer player) {
-            org.spigotmc.AsyncCatcher.catchOp("player tracker clear"); // Spigot
             if (this.seenBy.remove(player.connection)) {
                 this.serverEntity.removePairing(player);
                 if (this.seenBy.isEmpty()) {
                     ChunkMap.this.level.debugSynchronizers().dropEntity(this.entity);
                 }
             }
+            this.seenByUpdated(); // Purpur - async entity tracker (DivineMC)
         }
 
         public void updatePlayer(final ServerPlayer player) {
-            org.spigotmc.AsyncCatcher.catchOp("player tracker update"); // Spigot
             if (player != this.entity) {
+                if (org.bxteam.divinemc.config.DivineConfig.AsyncCategory.multithreadedEnabled && player == null) return; // Purpur - async entity tracker (DivineMC)
                 // Paper start - remove allocation of Vec3D here
                 // Vec3 deltaToPlayer = player.position().subtract(this.entity.position());
                 double deltaToPlayerX = player.getX() - this.entity.getX();
@@ -1382,6 +1516,7 @@ public class ChunkMap extends SimpleRegionStorage implements ChunkHolder.PlayerP
                 // CraftBukkit end
                 if (visibleToPlayer) {
                     if (this.seenBy.add(player.connection)) {
+                        this.seenByUpdated(); // Purpur - async entity tracker (DivineMC)
                         // Paper start - entity tracking events
                         if (io.papermc.paper.event.player.PlayerTrackEntityEvent.getHandlerList().getRegisteredListeners().length == 0 || new io.papermc.paper.event.player.PlayerTrackEntityEvent(player.getBukkitEntity(), this.entity.getBukkitEntity()).callEvent()) {
                         this.serverEntity.addPairing(player);
@@ -1395,6 +1530,7 @@ public class ChunkMap extends SimpleRegionStorage implements ChunkHolder.PlayerP
                         this.serverEntity.onPlayerAdd(); // Paper - fix desync when a player is added to the tracker
                     }
                 } else {
+                    this.seenByUpdated(); // Purpur - async entity tracker (DivineMC)
                     this.removePlayer(player);
                 }
             }
diff --git a/net/minecraft/server/level/ServerBossEvent.java b/net/minecraft/server/level/ServerBossEvent.java
index 3b996666ff183ba8f8a48306af7d2f9272eb30bf..c3946cd99867646c9cde9e7c6911550e78f2d157 100644
--- a/net/minecraft/server/level/ServerBossEvent.java
+++ b/net/minecraft/server/level/ServerBossEvent.java
@@ -13,7 +13,11 @@ import net.minecraft.network.protocol.game.ClientboundBossEventPacket;
 import net.minecraft.world.BossEvent;
 
 public class ServerBossEvent extends BossEvent {
-    private final Set<ServerPlayer> players = Sets.newHashSet();
+    // Purpur start - async entity tracker - players can be removed in async tracking (DivineMC)
+    private final Set<ServerPlayer> players = org.bxteam.divinemc.config.DivineConfig.AsyncCategory.multithreadedEnabled
+            ? Sets.newConcurrentHashSet()
+            : Sets.newHashSet();
+    // Purpur end - async entity tracker (DivineMC)
     private final Set<ServerPlayer> unmodifiablePlayers = Collections.unmodifiableSet(this.players);
     public boolean visible = true;
 
diff --git a/net/minecraft/server/level/ServerEntity.java b/net/minecraft/server/level/ServerEntity.java
index 5fad52157fc22755196ce9eaddaa2bc23c5bb86b..61fb663cc5d9871be5a8ef384f6ba31b19d484c7 100644
--- a/net/minecraft/server/level/ServerEntity.java
+++ b/net/minecraft/server/level/ServerEntity.java
@@ -123,7 +123,7 @@ public class ServerEntity {
                 MapId id = frame.cachedMapId; // Paper - Perf: Cache map ids on item frames
                 MapItemSavedData data = MapItem.getSavedData(id, this.level);
                 if (data != null) {
-                    for (final net.minecraft.server.network.ServerPlayerConnection connection : this.trackedPlayers) { // Paper
+                    for (final net.minecraft.server.network.ServerPlayerConnection connection : this.trackedPlayers.toArray(ChunkMap.TrackedEntity.EMPTY_OBJECT_ARRAY)) { // Paper // Purpur - async entity tracker (DivineMC)
                         final ServerPlayer player = connection.getPlayer(); // Paper
                         data.tickCarriedBy(player, itemStack, frame);
                         Packet<?> packet = data.getUpdatePacket(id, player);
@@ -406,8 +406,6 @@ public class ServerEntity {
                 // CraftBukkit end
                 this.synchronizer.sendToTrackingPlayersAndSelf(new ClientboundUpdateAttributesPacket(this.entity.getId(), attributes));
             }
-
-            attributes.clear();
         }
     }
 
diff --git a/net/minecraft/server/level/ServerLevel.java b/net/minecraft/server/level/ServerLevel.java
index e832db7e1493a47a05549d40172bd13e1626f6a9..d156b872ed38f19df15290f07ab1b3866c14574f 100644
--- a/net/minecraft/server/level/ServerLevel.java
+++ b/net/minecraft/server/level/ServerLevel.java
@@ -2674,7 +2674,6 @@ public class ServerLevel extends Level implements WorldGenLevel, ServerEntityGet
 
     @Override
     public LevelEntityGetter<Entity> getEntities() {
-        org.spigotmc.AsyncCatcher.catchOp("Chunk getEntities call"); // Spigot
         return this.moonrise$getEntityLookup(); // Paper - rewrite chunk system
     }
 
@@ -2980,7 +2979,7 @@ public class ServerLevel extends Level implements WorldGenLevel, ServerEntityGet
                         }
 
                         map.carriedByPlayers.remove(player);
-                        if (map.carriedBy.removeIf(holdingPlayer -> holdingPlayer.player == player)) {
+                        if (map.carriedBy.removeIf(holdingPlayer -> holdingPlayer != null && holdingPlayer.player == player)) { // Purpur - async entity tracker (DivineMC)
                             map.decorations.remove(player.getName().getString());
                         }
                     }
diff --git a/net/minecraft/server/network/ServerGamePacketListenerImpl.java b/net/minecraft/server/network/ServerGamePacketListenerImpl.java
index 3671945aa5d94052629c97dfbf6de581c9b20d04..68bcb8dff472c06901e0286158a9f57fac277d1e 100644
--- a/net/minecraft/server/network/ServerGamePacketListenerImpl.java
+++ b/net/minecraft/server/network/ServerGamePacketListenerImpl.java
@@ -2001,7 +2001,6 @@ public class ServerGamePacketListenerImpl
     }
 
     public void internalTeleport(PositionMoveRotation destination, Set<Relative> relatives) {
-        org.spigotmc.AsyncCatcher.catchOp("teleport"); // Paper
         // Paper start - Prevent teleporting dead entities
         if (this.player.isRemoved()) {
             LOGGER.info("Attempt to teleport removed player {} restricted", this.player.getScoreboardName());
diff --git a/net/minecraft/world/entity/LivingEntity.java b/net/minecraft/world/entity/LivingEntity.java
index dc56736203d68619f97ba2f4f2716b41bdeff714..1c6c9a6f4cc5919ceb0396a5a2262ff6beca396f 100644
--- a/net/minecraft/world/entity/LivingEntity.java
+++ b/net/minecraft/world/entity/LivingEntity.java
@@ -1356,13 +1356,13 @@ public abstract class LivingEntity extends Entity implements Attackable, Waypoin
     }
 
     private void refreshDirtyAttributes() {
-        Set<AttributeInstance> attributesToUpdate = this.getAttributes().getAttributesToUpdate();
+        // Purpur start - async entity tracker (DivineMC)
+        int[] attributesToUpdate = this.getAttributes().getAttributesToUpdateIds();
 
-        for (AttributeInstance changedAttributeInstance : attributesToUpdate) {
-            this.onAttributeUpdated(changedAttributeInstance.getAttribute());
+        for (int attribute : attributesToUpdate) {
+            this.onAttributeUpdated(net.minecraft.core.registries.BuiltInRegistries.ATTRIBUTE.get(attribute).orElseThrow());
         }
-
-        attributesToUpdate.clear();
+        // Purpur end - async entity tracker (DivineMC)
     }
 
     protected void onAttributeUpdated(final Holder<Attribute> attribute) {
diff --git a/net/minecraft/world/entity/ai/attributes/Attribute.java b/net/minecraft/world/entity/ai/attributes/Attribute.java
index ad3dc7f69eb6e5a167118c4fe16c2fe4083522dd..f1d848cbe236b8643ef5d009a771bd267685341c 100644
--- a/net/minecraft/world/entity/ai/attributes/Attribute.java
+++ b/net/minecraft/world/entity/ai/attributes/Attribute.java
@@ -16,10 +16,14 @@ public class Attribute {
     private boolean syncable;
     private final String descriptionId;
     public Attribute.Sentiment sentiment = Attribute.Sentiment.POSITIVE;
-
+    // Purpur start - async entity tracker (DivineMC)
+    public final int uid;
+    private static final java.util.concurrent.atomic.AtomicInteger SIZE = new java.util.concurrent.atomic.AtomicInteger();
+    // Purpur end - async entity tracker (DivineMC)
     protected Attribute(final String descriptionId, final double defaultValue) {
         this.defaultValue = defaultValue;
         this.descriptionId = descriptionId;
+        this.uid = SIZE.getAndAdd(1); // Purpur - async entity tracker (DivineMC)
     }
 
     public double getDefaultValue() {
diff --git a/net/minecraft/world/entity/ai/attributes/AttributeInstance.java b/net/minecraft/world/entity/ai/attributes/AttributeInstance.java
index 0ed3fea287c65da82a8823d1949828b0c55d29b2..e0c76cf83b6f20223001f088dc3ef27940a128ba 100644
--- a/net/minecraft/world/entity/ai/attributes/AttributeInstance.java
+++ b/net/minecraft/world/entity/ai/attributes/AttributeInstance.java
@@ -20,8 +20,24 @@ import org.jspecify.annotations.Nullable;
 public class AttributeInstance {
     private final Holder<Attribute> attribute;
     private final Map<AttributeModifier.Operation, Map<Identifier, AttributeModifier>> modifiersByOperation = Maps.newEnumMap(AttributeModifier.Operation.class);
-    private final Map<Identifier, AttributeModifier> modifierById = new Object2ObjectArrayMap<>();
-    private final Map<Identifier, AttributeModifier> permanentModifiers = new Object2ObjectArrayMap<>();
+    // Purpur start - async entity tracker (DivineMC)
+    private final Map<Identifier, AttributeModifier> modifierById;
+    {
+        if (org.bxteam.divinemc.config.DivineConfig.AsyncCategory.multithreadedEnabled) {
+            modifierById = it.unimi.dsi.fastutil.objects.Object2ObjectMaps.synchronize(new Object2ObjectArrayMap<>(), this);
+        } else {
+            modifierById = new Object2ObjectArrayMap<>();
+        }
+    }
+    private final Map<Identifier, AttributeModifier> permanentModifiers;
+    {
+        if (org.bxteam.divinemc.config.DivineConfig.AsyncCategory.multithreadedEnabled) {
+            permanentModifiers = it.unimi.dsi.fastutil.objects.Object2ObjectMaps.synchronize(new Object2ObjectArrayMap<>(), this);
+        } else {
+            permanentModifiers = new Object2ObjectArrayMap<>();
+        }
+    }
+    // Purpur end - async entity tracker (DivineMC)
     private double baseValue;
     private boolean dirty = true;
     private double cachedValue;
@@ -50,7 +66,13 @@ public class AttributeInstance {
 
     @VisibleForTesting
     Map<Identifier, AttributeModifier> getModifiers(final AttributeModifier.Operation operation) {
-        return this.modifiersByOperation.computeIfAbsent(operation, key -> new Object2ObjectOpenHashMap<>());
+        // Purpur start - async entity tracker (DivineMC)
+        return this.modifiersByOperation.computeIfAbsent(operation, operation1 -> {
+            if (org.bxteam.divinemc.config.DivineConfig.AsyncCategory.multithreadedEnabled)
+                return it.unimi.dsi.fastutil.objects.Object2ObjectMaps.synchronize(new Object2ObjectArrayMap<>(), this);
+            else return new Object2ObjectArrayMap<>();
+        });
+        // Purpur end - async entity tracker (DivineMC)
     }
 
     public Set<AttributeModifier> getModifiers() {
@@ -137,8 +159,12 @@ public class AttributeInstance {
 
     public double getValue() {
         if (this.dirty) {
-            this.cachedValue = this.calculateValue();
+            // Purpur start - async entity tracker (DivineMC)
+            double value = this.calculateValue();
+            this.cachedValue = value;
             this.dirty = false;
+            return value;
+            // Purpur end - async entity tracker (DivineMC)
         }
 
         return this.cachedValue;
@@ -184,7 +210,15 @@ public class AttributeInstance {
     }
 
     public AttributeInstance.Packed pack() {
-        return new AttributeInstance.Packed(this.attribute, this.baseValue, List.copyOf(this.permanentModifiers.values()));
+        // Purpur start - async entity tracker (DivineMC)
+        if (org.bxteam.divinemc.config.DivineConfig.AsyncCategory.multithreadedEnabled) {
+            synchronized (this) {
+                return new AttributeInstance.Packed(this.attribute, this.baseValue, List.copyOf(this.permanentModifiers.values()));
+            }
+        } else {
+            return new AttributeInstance.Packed(this.attribute, this.baseValue, List.copyOf(this.permanentModifiers.values()));
+        }
+        // Purpur end - async entity tracker (DivineMC)
     }
 
     public void apply(final AttributeInstance.Packed packed) {
diff --git a/net/minecraft/world/entity/ai/attributes/AttributeMap.java b/net/minecraft/world/entity/ai/attributes/AttributeMap.java
index fe5f4f79e5a1b091d15fa3bfd876b831e493655c..0c17c1d448cc4e2fcb1086f6944ac7b6a79b78e6 100644
--- a/net/minecraft/world/entity/ai/attributes/AttributeMap.java
+++ b/net/minecraft/world/entity/ai/attributes/AttributeMap.java
@@ -14,9 +14,11 @@ import net.minecraft.resources.Identifier;
 import org.jspecify.annotations.Nullable;
 
 public class AttributeMap {
-    private final Map<Holder<Attribute>, AttributeInstance> attributes = new Object2ObjectOpenHashMap<>();
-    private final Set<AttributeInstance> attributesToSync = new ObjectOpenHashSet<>();
-    private final Set<AttributeInstance> attributesToUpdate = new ObjectOpenHashSet<>();
+    // Purpur start - async entity tracker (DivineMC)
+    private final Map<Holder<Attribute>, AttributeInstance> attributes = new org.bxteam.divinemc.util.map.AttributeInstanceArrayMap();
+    private final org.bxteam.divinemc.util.map.AttributeInstanceSet attributesToSync =  new org.bxteam.divinemc.util.map.AttributeInstanceSet((org.bxteam.divinemc.util.map.AttributeInstanceArrayMap) attributes);
+    private final org.bxteam.divinemc.util.map.AttributeInstanceSet attributesToUpdate = new org.bxteam.divinemc.util.map.AttributeInstanceSet((org.bxteam.divinemc.util.map.AttributeInstanceArrayMap) attributes);
+    // Purpur end - async entity tracker (DivineMC)
     private final AttributeSupplier supplier;
     private final net.minecraft.world.entity.LivingEntity entity; // Purpur - Ridables
 
@@ -30,27 +32,50 @@ public class AttributeMap {
         this.supplier = defaultAttributes;
     }
 
-    private void onAttributeModified(final AttributeInstance attributeInstance) {
+    private synchronized void onAttributeModified(final AttributeInstance attributeInstance) { // Purpur - async entity tracker (DivineMC)
         this.attributesToUpdate.add(attributeInstance);
         if (attributeInstance.getAttribute().value().isClientSyncable() && (entity == null || entity.shouldSendAttribute(attributeInstance.getAttribute().value()))) { // Purpur - Ridables
             this.attributesToSync.add(attributeInstance);
         }
     }
 
-    public Set<AttributeInstance> getAttributesToSync() {
-        return this.attributesToSync;
+    // Purpur start - async entity tracker (DivineMC)
+    private static final AttributeInstance[] EMPTY_ATTRIBUTE_INSTANCE = new AttributeInstance[0];
+    public synchronized Set<AttributeInstance> getAttributesToSync() {
+        var clone = it.unimi.dsi.fastutil.objects.ReferenceArraySet.ofUnchecked(attributesToSync.toArray(EMPTY_ATTRIBUTE_INSTANCE));
+        this.attributesToSync.clear();
+        return clone;
     }
 
-    public Set<AttributeInstance> getAttributesToUpdate() {
-        return this.attributesToUpdate;
+    public synchronized Set<AttributeInstance> getAttributesToUpdate() {
+        var clone = it.unimi.dsi.fastutil.objects.ReferenceArraySet.ofUnchecked(attributesToUpdate.toArray(EMPTY_ATTRIBUTE_INSTANCE));
+        this.attributesToUpdate.clear();
+        return clone;
     }
 
+    public synchronized int[] getAttributesToUpdateIds() {
+        int[] clone = attributesToUpdate.inner.toIntArray();
+        this.attributesToUpdate.clear();
+        return clone;
+    }
+    // Purpur end - async entity tracker (DivineMC)
     public Collection<AttributeInstance> getSyncableAttributes() {
         return this.attributes.values().stream().filter(instance -> instance.getAttribute().value().isClientSyncable() && (entity == null || entity.shouldSendAttribute(instance.getAttribute().value()))).collect(Collectors.toList()); // Purpur - Ridables
     }
 
     public @Nullable AttributeInstance getInstance(final Holder<Attribute> attribute) {
-        return this.attributes.computeIfAbsent(attribute, key -> this.supplier.createInstance(this::onAttributeModified, (Holder<Attribute>)key));
+        // Purpur start - async entity tracker (DivineMC)
+        AttributeInstance attributeInstance;
+        if ((attributeInstance = this.attributes.get(attribute)) == null) {
+            AttributeInstance newValue;
+            if ((newValue = this.supplier.createInstance(this::onAttributeModified, attribute)) != null) {
+                attributes.put(attribute, newValue);
+                return newValue;
+            }
+        }
+
+        return attributeInstance;
+        // Purpur end - async entity tracker (DivineMC)
     }
 
     public boolean hasAttribute(final Holder<Attribute> attribute) {
diff --git a/net/minecraft/world/entity/ai/attributes/AttributeSupplier.java b/net/minecraft/world/entity/ai/attributes/AttributeSupplier.java
index 5d2c1cfbdd20c5c195790b5241c3a607c51aa72c..e1784546d883b386319e10f93e3716ffedc05916 100644
--- a/net/minecraft/world/entity/ai/attributes/AttributeSupplier.java
+++ b/net/minecraft/world/entity/ai/attributes/AttributeSupplier.java
@@ -11,7 +11,7 @@ public class AttributeSupplier {
     private final Map<Holder<Attribute>, AttributeInstance> instances;
 
     private AttributeSupplier(final Map<Holder<Attribute>, AttributeInstance> instances) {
-        this.instances = instances;
+        this.instances = new org.bxteam.divinemc.util.map.AttributeInstanceArrayMap(instances); // Purpur - async entity tracker (DivineMC)
     }
 
     public AttributeInstance getAttributeInstance(final Holder<Attribute> attribute) {
diff --git a/net/minecraft/world/entity/vehicle/minecart/NewMinecartBehavior.java b/net/minecraft/world/entity/vehicle/minecart/NewMinecartBehavior.java
index 4a8bbb06e0cbb8974fa4f2e509eafe38b7d38173..00b98c33c6eebc702f59fe1e1cc9c3dd20b8c606 100644
--- a/net/minecraft/world/entity/vehicle/minecart/NewMinecartBehavior.java
+++ b/net/minecraft/world/entity/vehicle/minecart/NewMinecartBehavior.java
@@ -34,13 +34,20 @@ public class NewMinecartBehavior extends MinecartBehavior {
     private int cachedLerpDelay;
     private float cachedPartialTick;
     private int lerpDelay = 0;
-    public final List<NewMinecartBehavior.MinecartStep> lerpSteps = new LinkedList<>();
+    public final List<NewMinecartBehavior.MinecartStep> lerpSteps; // Purpur - async entity tracker (DivineMC)
     public final List<NewMinecartBehavior.MinecartStep> currentLerpSteps = new LinkedList<>();
     public double currentLerpStepsTotalWeight = 0.0;
     public NewMinecartBehavior.MinecartStep oldLerp = NewMinecartBehavior.MinecartStep.ZERO;
 
     public NewMinecartBehavior(final AbstractMinecart minecart) {
         super(minecart);
+        // Purpur start - async entity tracker (DivineMC)
+        if (org.bxteam.divinemc.config.DivineConfig.AsyncCategory.multithreadedEnabled) {
+            this.lerpSteps = it.unimi.dsi.fastutil.objects.ObjectLists.synchronize(new it.unimi.dsi.fastutil.objects.ObjectArrayList<>());
+        } else {
+            this.lerpSteps = new it.unimi.dsi.fastutil.objects.ObjectArrayList<>();
+        }
+        // Purpur end - async entity tracker (DivineMC)
     }
 
     @Override
diff --git a/net/minecraft/world/level/saveddata/maps/MapItemSavedData.java b/net/minecraft/world/level/saveddata/maps/MapItemSavedData.java
index 54f30879b55e6b9557f8956652076d1d3b7dff53..925978f86b9a7014ae5085ef19e1fc5196e0dde2 100644
--- a/net/minecraft/world/level/saveddata/maps/MapItemSavedData.java
+++ b/net/minecraft/world/level/saveddata/maps/MapItemSavedData.java
@@ -215,6 +215,7 @@ public class MapItemSavedData extends SavedData {
 
         for (int i = 0; i < this.carriedBy.size(); i++) {
             MapItemSavedData.HoldingPlayer otherHoldingPlayer = this.carriedBy.get(i);
+            if (otherHoldingPlayer == null) continue; // Purpur - async entity tracker (DivineMC)
             Player otherPlayer = otherHoldingPlayer.player;
             String otherPlayerName = otherPlayer.getPlainTextName();
             if (!otherPlayer.isRemoved() && (placedInFrame != null || otherPlayer.getInventory().contains(mapMatcher))) {
