From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Carlo=20Maria=20Card=C3=AD?=
 <112445885+MathsAnalysis@users.noreply.github.com>
Date: Thu, 4 Jun 2026 09:35:46 +0200
Subject: [PATCH] MSPT Tracking for each world


diff --git a/net/minecraft/server/MinecraftServer.java b/net/minecraft/server/MinecraftServer.java
index 0c9c22d23b0df2da1f2f366589c4e6d001a37c87..8e7f14220ebd2e6d0fddb020d7ec8a07c1f05ae7 100644
--- a/net/minecraft/server/MinecraftServer.java
+++ b/net/minecraft/server/MinecraftServer.java
@@ -1886,6 +1886,8 @@ public abstract class MinecraftServer extends ReentrantBlockableEventLoop<TickTa
             profiler.push(() -> level + " " + level.dimension().identifier());
             profiler.push("tick");
 
+            // Purpur start - per-world MSPT tracking (DivineMC)
+            final long worldTickStart = Util.getNanos();
             try {
                 level.tick(haveTime);
             } catch (Throwable t) {
@@ -1893,6 +1895,11 @@ public abstract class MinecraftServer extends ReentrantBlockableEventLoop<TickTa
                 level.fillReportDetails(report);
                 throw new ReportedException(report);
             }
+            final long worldTickDuration = Util.getNanos() - worldTickStart;
+            level.tickTimes5s.add(this.tickCount, worldTickDuration);
+            level.tickTimes10s.add(this.tickCount, worldTickDuration);
+            level.tickTimes60s.add(this.tickCount, worldTickDuration);
+            // Purpur end - per-world MSPT tracking (DivineMC)
 
             profiler.pop();
             profiler.pop();
