From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Carlo=20Maria=20Card=C3=AD?=
 <112445885+MathsAnalysis@users.noreply.github.com>
Date: Sat, 6 Jun 2026 01:31:54 +0200
Subject: [PATCH] Raytrace AntiXray movement reveal tick hook


diff --git a/net/minecraft/server/MinecraftServer.java b/net/minecraft/server/MinecraftServer.java
index 6f5eab24b881de14373d24a8a026d6bfb0a5e43c..bfb89005974221e97118718f54fb5882ebb296bd 100644
--- a/net/minecraft/server/MinecraftServer.java
+++ b/net/minecraft/server/MinecraftServer.java
@@ -1699,6 +1699,7 @@ public abstract class MinecraftServer extends ReentrantBlockableEventLoop<TickTa
         this.tickCount++;
         this.tickRateManager.tick();
         this.tickChildren(haveTime);
+        org.bxteam.divinemc.antixray.RaytraceAntiXrayEngine.tickServer(); // DivineMC - raytrace anti-xray movement reveal pass
         if (nano - this.lastServerStatus >= STATUS_EXPIRE_TIME_NANOS) {
             this.lastServerStatus = nano;
             this.status = this.buildServerStatus();
